### PR TITLE
ACS-2321 Downloads throws 412 error if content is archived

### PR DESCRIPTION
**Pull request by PunchedChimera**

### CODE REVIEW

File: remote-api/src/main/java/org/alfresco/rest/api/impl/DownloadsImpl.java
Comment (montgolfiere) [inline remote-api/src/main/java/org/alfresco/rest/api/impl/DownloadsImpl.java:103]: Add discovery - only makes sense to call if either S3 Connector (or Azure Connector in future) are deployed.

**Review by montgolfiere — COMMENTED** [review]
I've added a comment about only running if S3 Connector or Azure Connector are deployed.

More generally, this request (to create a "download" operation) could take some time if the folder hierarchies are deep - ie. large number of content nodes to process. 

Maybe we need a configurable cut-off, eg. after first 500 content items.

At that point, defer to background action and then ensure if ArchiveException occurs that this is caught and handled appropriately, ie. marking the download as failed, with reason/error code (if possible).

File: remote-api/src/main/java/org/alfresco/rest/api/impl/DownloadsImpl.java
Comment (PunchedChimera) [reply remote-api/src/main/java/org/alfresco/rest/api/impl/DownloadsImpl.java]: Thanks, I've added the discovery using the `ModuleService` and add a hardcoded limit of 500 to the number of checks.

File: remote-api/src/main/java/org/alfresco/rest/api/impl/DownloadsImpl.java
Comment (montgolfiere) [inline remote-api/src/main/java/org/alfresco/rest/api/impl/DownloadsImpl.java:251]: How likely (when) will a node appear more than once (given we're only walking primary child assocs) ? 

Could we use two HashSet instead (eg. folderRefs and nonFolderRefs) and then only track a total count ?

File: remote-api/src/main/java/org/alfresco/rest/api/impl/DownloadsImpl.java
Comment (montgolfiere) [inline remote-api/src/main/java/org/alfresco/rest/api/impl/DownloadsImpl.java:285]: Do we need this extra method (could we move up into protected "checkArchiveStatus" ?

File: remote-api/src/main/java/org/alfresco/rest/api/impl/DownloadsImpl.java
Comment (montgolfiere) [inline remote-api/src/main/java/org/alfresco/rest/api/impl/DownloadsImpl.java:114]: Magic number - at minimum should be a constant. We should consider make it a configurable property (ie. add to "repository.properties" and inject into this bean.

**Review by montgolfiere — COMMENTED** [review]
I've added a few minor comments.

We should also log in JIRA separate tech debt to improve the handling of errors that might occur for larger downloads (eg. if beyond our initial limit when accepting the request).

**Review by KikoUK — APPROVED** [review]
LGTM once other comments are addressed.

File: remote-api/src/main/java/org/alfresco/rest/api/impl/DownloadsImpl.java
Comment (PunchedChimera) [reply remote-api/src/main/java/org/alfresco/rest/api/impl/DownloadsImpl.java:239]: Yes, I agree. I'm adding to the next commit.

File: remote-api/src/main/java/org/alfresco/rest/api/impl/DownloadsImpl.java
Comment (PunchedChimera) [reply remote-api/src/main/java/org/alfresco/rest/api/impl/DownloadsImpl.java:251]: It's unlikely but possible. Image the following, simple, file heirarchy: folder1/folder2/content. The following download request is legal: 
{  
    "nodeIds": [    
        "node for folder1",
        "node for folder2",
        "node for content
    ]
}
You can see that we could attempt to look at the child refs of folder2 and check content multiple times.
I have attempted to replicate this complexity in the unit tests provided.

File: remote-api/src/main/java/org/alfresco/rest/api/impl/DownloadsImpl.java
Comment (PunchedChimera) [reply remote-api/src/main/java/org/alfresco/rest/api/impl/DownloadsImpl.java:285]: Doesn't look like it, I'll move in the next commit. I think I was keep my options open for unit tests but never used it.

File: remote-api/src/main/java/org/alfresco/rest/api/impl/DownloadsImpl.java
Comment (PunchedChimera) [reply remote-api/src/main/java/org/alfresco/rest/api/impl/DownloadsImpl.java:114]: Thanks, refactoring to global property to be overriden by the connectors

Comment (PunchedChimera) [reply remote-api/src/main/java/org/alfresco/rest/api/impl/DownloadsImpl.java:114]: Thanks, refactoring to global property to be overriden by the connectors

File: remote-api/src/main/java/org/alfresco/rest/api/impl/DownloadsImpl.java
Comment (mpichura) [inline remote-api/src/main/java/org/alfresco/rest/api/impl/DownloadsImpl.java]: I'm happy with these string literals (and `moduleServiceContains` solution) but would it be possible to either move them to constants or even do something like:
```
public static final String [] CLOUD_CONNECTOR_MODULES = {"org_alfresco_integrations_AzureConnector", "org_alfresco_integrations_S3Connector"};
@Experimental
protected boolean canCheckArchived() {
    return Arrays.stream(CLOUD_CONNECTOR_MODULES).anyMatch(m-> moduleService.getModule(m) != null);
}
```

File: remote-api/src/main/java/org/alfresco/rest/api/impl/DownloadsImpl.java
Comment (montgolfiere) [reply remote-api/src/main/java/org/alfresco/rest/api/impl/DownloadsImpl.java:260]: Ah ok - thanks.

File: remote-api/src/main/java/org/alfresco/rest/api/impl/DownloadsImpl.java
Comment (montgolfiere) [reply remote-api/src/main/java/org/alfresco/rest/api/impl/DownloadsImpl.java]: Thanks. I'm not sure connectors need to override (do they) ? Wouldn't a global (ie. system-wide) be sufficient ?

**Review by mpichura — APPROVED** [review]
Some suggestions which not necessarily need to be considered in this iteration.

File: remote-api/src/test/java/org/alfresco/rest/api/impl/DownloadsImplUnitTest.java
Comment (PunchedChimera) [reply remote-api/src/test/java/org/alfresco/rest/api/impl/DownloadsImplUnitTest.java]: Refactored and fixed the tests 👍

File: remote-api/src/main/java/org/alfresco/rest/api/impl/DownloadsImpl.java
Comment (PunchedChimera) [reply remote-api/src/main/java/org/alfresco/rest/api/impl/DownloadsImpl.java]: Love this, added 👍 
We could easily, make the CLOUD_CONNECTOR_MODULES a config bean later if needed.

File: remote-api/src/main/java/org/alfresco/rest/api/impl/DownloadsImpl.java
Comment (PunchedChimera) [reply remote-api/src/main/java/org/alfresco/rest/api/impl/DownloadsImpl.java]: I have since refactor and re-added this method. The new implementation recommended by @mpichura is better sat in it's own method.

**Review by montgolfiere — APPROVED** [review]
Thanks for the updates. 

Do we need jira reminder to review cleanup of pending downloads (eg. if error occurs midway) ?

**Pull request merged**
